crates/swc_common/src/plugin.rs

```diff
@@ -6,40 +6,66 @@
 
 use crate::{syntax_pos::Mark, SyntaxContext};
 use anyhow::Error;
-use std::any::type_name;
+use std::{any::type_name, marker::PhantomData};
 
-/// Serialize ast to pass into plugins. This is being called from host side.
-/// TODO: https://github.com/swc-project/swc/issues/3167
-#[cfg(feature = "plugin-base")]
-pub fn serialize_for_plugin<T>(t: &T) -> Result<rkyv::AlignedVec, Error>
-where
-    T: rkyv::Serialize<rkyv::ser::serializers::AllocSerializer<512>>,
-{
-    rkyv::to_bytes::<_, 512>(t).map_err(|err| match err {
-        rkyv::ser::serializers::CompositeSerializerError::SerializerError(e) => e.into(),
-        rkyv::ser::serializers::CompositeSerializerError::ScratchSpaceError(e) => {
-            Error::msg("AllocScratchError")
-        }
-        rkyv::ser::serializers::CompositeSerializerError::SharedError(e) => {
-            Error::msg("SharedSerializeMapError")
-        }
-    })
+/// Wraps internal representation of serialized data. Consumers should not
+/// rely on specific details of byte format struct contains: it is
+/// strictly implementation detail which can change anytime.
+pub struct Serialized<T> {
+    field: T,
 }
 
-/// TODO: https://github.com/swc-project/swc/issues/3167
-#[cfg(feature = "plugin-mode")]
-pub fn deserialize_for_plugin<T>(bytes: &[u8]) -> Result<T, Error>
-where
-    T: rkyv::Archive
-        + rkyv::with::DeserializeWith<<T as rkyv::Archive>::Archived, T, rkyv::Infallible>,
-{
-    use anyhow::Context;
-    use rkyv::Deserialize;
-
-    let archived = unsafe { rkyv::archived_root::<T>(&bytes[..]) };
-
-    archived
-        .deserialize(&mut rkyv::Infallible)
-        .map(|r: rkyv::with::With<T, T>| r.into_inner())
-        .with_context(|| format!("failed to deserialize `{}`", type_name::<T>()))
+pub type PluginSerializedByte = Serialized<rkyv::AlignedVec>;
+
+#[cfg(feature = "plugin-base")]
+impl Serialized<rkyv::AlignedVec> {
```

Comment:
@kdy1 thoughts on this? I don't feel this is elegant enough, but this seems could be a way to make `Serialized` struct to be generic across different usecases later.

Reply:
I think this is fine.

Reply:
Oh, sorry. I misread the code.

`Serialized<rkyv::AlignedVec>` seems weird to me, as it's not a serialized form of `rkyv::AlignedVec`.

Reply:
Yes, that's the part I'm feeling bit.. awkward.

Reply:
in here T represents internal field type, not the type of original type to be deserialized into. I presume there should be some better way?

Reply:
ok, removed generic to simplify things, based on discussion https://github.com/swc-project/swc/pull/3227#discussion_r781868216 that we'll rely on rkyv in most of cases for the serialization.

But also, field is hidden and there's no public direct access to rkyv from the consumers, we may not need worry too much for the breaking changes.